consensus/src/pipeline/execution_client.rs

```diff
@@ -0,0 +1,359 @@
+// Copyright © Aptos Foundation
+// Parts of the project are originally copyright © Meta Platforms, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    counters,
+    error::StateSyncError,
+    network::{IncomingCommitRequest, NetworkSender},
+    network_interface::{ConsensusMsg, ConsensusNetworkClient},
+    payload_manager::PayloadManager,
+    pipeline::{
+        buffer_manager::{OrderedBlocks, ResetAck, ResetRequest, ResetSignal},
+        decoupled_execution_utils::prepare_phases_and_buffer_manager,
+        errors::Error,
+        signing_phase::CommitSignerProvider,
+    },
+    state_computer::ExecutionProxy,
+    state_replication::{StateComputer, StateComputerCommitCallBackType},
+    transaction_deduper::create_transaction_deduper,
+    transaction_shuffler::create_transaction_shuffler,
+};
+use anyhow::Result;
+use aptos_bounded_executor::BoundedExecutor;
+use aptos_channels::{aptos_channel, message_queues::QueueStyle};
+use aptos_consensus_types::{common::Author, pipelined_block::PipelinedBlock};
+use aptos_executor_types::ExecutorResult;
+use aptos_logger::prelude::*;
+use aptos_network::{application::interface::NetworkClient, protocols::network::Event};
+use aptos_types::{
+    epoch_state::EpochState,
+    ledger_info::LedgerInfoWithSignatures,
+    on_chain_config::{FeatureFlag, Features, OnChainExecutionConfig},
+};
+use fail::fail_point;
+use futures::{
+    channel::{mpsc::UnboundedSender, oneshot},
+    SinkExt,
+};
+use futures_channel::mpsc::unbounded;
+use move_core_types::account_address::AccountAddress;
+use std::sync::{Arc, RwLock};
+
+#[async_trait::async_trait]
+pub trait ExecutionClient: Send + Sync {
+    /// Initialize the execution phase for a new epoch.
+    async fn init_for_new_epoch(
```

Comment:
@bchocho - I agree, this is not so clean but initializing a new `ExecutionClient` looks more ugly - especially because I need to pass weird `Arc<RWLock<ExecutionClient>>` passed around.  Currently its easy to reason about the locking semantics as the lock is contained in the `ExecutionClient`.  I will add a follow up to see if I can clean up the interface.

Reply:
Yeah let's see how this evolves as we iterate